Deploy and manage: Autoscaling migration #344 (#606)
elastic/docs-content: deploy-manage/autoscaling.md

Reviewer: because this is the overview page and TM autoscaling (a thing available to serverless) is under it, we might want to make this page apply to serverless as well and just make what's available to each deployment type really clear.

:::{{tip}} - Serverless handles autoscaling for you
By default, {{serverless-full}} automatically scales your {{es}} resources based on your usage. You don't need to enable autoscaling.
:::

Trained model autoscaling is available on all cloud deployments, and {{serverless-short}}.

A policy is composed of a list of roles and a list of deciders. The policy governs the nodes matching the roles. The deciders provide independent estimates of the capacity required. See [Autoscaling deciders](../deploy-manage/autoscaling/autoscaling-deciders.md) for details on available deciders.

Autoscaling supports:
* Scaling machine learning nodes up and down.
* Scaling data nodes up based on storage.
* Scaling trained model deployments

::::{note} 
Autoscaling is not supported on Debian 8.

> ::::

Author: Thanks for the feedback! I’ve updated the doc to separate regular autoscaling (cluster autoscaling) and trained model autoscaling. The general definition now applies to all deployment types, including serverless. Let me know if anything still needs tweaking!

  - https://www.elastic.co/guide/en/cloud/current/ec-autoscaling.html
  - https://www.elastic.co/guide/en/cloud-enterprise/current/ece-autoscaling.html
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/xpack-autoscaling.html
applies_to:
  deployment:
    ece: ga
    ess: ga
    eck: ga
---

> # Autoscaling

Reviewer: this page is tagged for ECK, but doesn't talk about ECK at all. to make sure that people can find docs relevant to them, we probably need to update the section structure:

* Autoscaling (just overviews about the concepts, and links to the other pages in the section)
* Autoscaling in ECE and ECH (ECE and ECH specific info, example, API instructions)
* Autoscaling in ECK (section for deployments autoscaling and other autoscaling)
* Autoscaling deciders
* Trained model autoscaling 

fyi, part of the confusion is my lazy mapping of these pages - sorry for that.

::::{note} 
{{cloud-only}}

> ::::

Reviewer: This note doesn't render properly. you need to edit the contents of the `{{cloud-only}}` var so it uses the markdown link formatting.

consider instead: 

```
::::{admonition} Indirect use only 
This feature is designed for indirect use by {{ech}}, {{ece}}, and {{eck}}. Direct use is not supported.
::::
```

> The autoscaling feature allows an operator to create tiers of nodes. These nodes monitor themselves to decide if they need to scale, based on an operator-defined policy. An Elasticsearch cluster can use the autoscaling API to report if it needs more resources to meet the policy. For example, an operator could define a policy that a warm tier should scale on available disk space. Elasticsearch monitors disk space in the warm tier. If it predicts low disk space for current and future shard copies, the autoscaling API will report that the cluster needs to scale. It remains the responsibility of the operator to add the additional resources that the cluster signals it requires.

Reviewer: autoscaling is a very attractive feature, and people might be confused as to why it's not offered for serverless. we should add a note to help people to understand how autoscaling works in serverless

```suggestion
The autoscaling feature allows an operator to create tiers of nodes. These nodes monitor themselves to decide if they need to scale, based on an operator-defined policy. An Elasticsearch cluster can use the autoscaling API to report if it needs more resources to meet the policy. For example, an operator could define a policy that a warm tier should scale on available disk space. Elasticsearch monitors disk space in the warm tier. If it predicts low disk space for current and future shard copies, the autoscaling API will report that the cluster needs to scale. It remains the responsibility of the operator to add the additional resources that the cluster signals it requires.

:::{{tip}} - Serverless handles autoscaling for you
By default, {{serverless-full}} automatically scales your {{es}} resources based on your usage. You don't need to enable autoscaling.
:::
```

    * To control the maximum size of your machine learning nodes and ensure they will not scale above a certain size, you can use the maximum size per zone field.
    * To control the minimum size of your machine learning nodes and ensure the autoscaling mechanism will not scale machine learning below a certain size, you can use the minimum size per zone field.
    * The determination of when to scale is based on the expected memory and CPU requirements for the currently configured machine learning jobs and trained models.

::::{note}
For any Elasticsearch component the number of availability zones is not affected by autoscaling. You can always set the number of availability zones manually and the autoscaling mechanism will add or remove capacity per availability zone.
::::

## When does autoscaling occur?[ec-autoscaling-factors]

$$$ece-autoscaling-factors$$$$$$ech-autoscaling-factors$$$Several factors determine when data tiers or machine learning nodes are scaled.

For a data tier, an autoscaling event can be triggered in the following cases:

* Based on an assessment of how shards are currently allocated, and the amount of storage and buffer space currently available.

> When past behavior on a hot tier indicates that the influx of data can increase significantly in the near future. Refer to [Reactive storage decider](../deploy-manage/autoscaling/autoscaling-deciders.md) and [Proactive storage decider](../deploy-manage/autoscaling/autoscaling-deciders.md) for more detail.

Reviewer: ```suggestion
* When past behavior on a hot tier indicates that the influx of data can increase significantly in the near future. Refer to [Reactive storage decider](../deploy-manage/autoscaling/autoscaling-deciders.md) and [Proactive storage decider](../deploy-manage/autoscaling/autoscaling-deciders.md) for more detail.
```

On machine learning nodes, scaling is determined by an estimate of the memory and CPU requirements for the currently configured jobs and trained models. When a new machine learning job tries to start, it looks for a node with adequate native memory and CPU capacity. If one cannot be found, it stays in an `opening` state. If this waiting job exceeds the queueing limit set in the machine learning decider, a scale up is requested. Conversely, as machine learning jobs run, their memory and CPU usage might decrease or other running jobs might finish or close. In this case, if the duration of decreased resource usage exceeds the set value for `down_scale_delay`, a scale down is requested. Check [Machine learning decider](../deploy-manage/autoscaling/autoscaling-deciders.md) for more detail. To learn more about machine learning jobs in general, check [Create anomaly detection jobs](../explore-analyze/machine-learning/anomaly-detection/ml-ad-run-jobs.md#ml-ad-create-job)

On a highly available deployment, autoscaling events are always applied to instances in each availability zone simultaneously, to ensure consistency.

## Notifications[ec-autoscaling-notifications]

> $$$ece-autoscaling-notifications$$$$$$ech-autoscaling-notifications$$$

Reviewer: let's try to kill these anchors and see if CI complains

> ::::{tab-set}
> 
> :::{tab-item} {{ech}} and {{ecloud}} - Heroku
> In the event that a data tier or machine learning node scales up to its maximum possible size, you’ll receive an email, and a notice also appears on the deployment overview page prompting you to adjust your autoscaling settings to ensure optimal performance.
> :::
> 
> :::{tab-item} {{ece}}
> In the event that a data tier or machine learning node scales up to its maximum possible size, a notice appears on the deployment overview page prompting you to adjust your autoscaling settings in order to ensure optimal performance.
> 
> A warning is also issued in the ECE `service-constructor` logs with the field `labels.autoscaling_notification_type` and a value of `data-tier-at-limit` (for a fully scaled data tier) or `ml-tier-at-limit` (for a fully scaled machine learning node). The warning is indexed in the `logging-and-metrics` deployment, so you can use that event to [configure an email notification](../explore-analyze/alerts-cases/watcher.md)
> :::
> 
> ::::

Reviewer: Couple of things here: 

* just throw out the heroku stuff. heroku is just elastic cloud, and we are removing all references to it.
* In the case where only one sentence differs, you can just add a note indicating the context:

```suggestion
In the event that a data tier or machine learning node scales up to its maximum possible size, you’ll receive an email, and a notice also appears on the deployment overview page prompting you to adjust your autoscaling settings to ensure optimal performance.

In {{ece}} deployments, a warning is also issued in the ECE `service-constructor` logs with the field `labels.autoscaling_notification_type` and a value of `data-tier-at-limit` (for a fully scaled data tier) or `ml-tier-at-limit` (for a fully scaled machine learning node). The warning is indexed in the `logging-and-metrics` deployment, so you can use that event to [configure an email notification](../explore-analyze/alerts-cases/watcher.md).
```

## Restrictions and limitations[ec-autoscaling-restrictions]

$$$ece-autoscaling-restrictions$$$$$$ech-autoscaling-restrictions$$$The following are known limitations and restrictions with autoscaling:

* Autoscaling will not run if the cluster is unhealthy or if the last Elasticsearch plan failed.

> 
> ::::{tab-set}
> 
> :::{tab-item} {{ech}}
> * Trial deployments cannot be configured to autoscale beyond the normal Trial deployment size limits. The maximum size per zone is increased automatically from the Trial limit when you convert to a paid subscription.
> * ELSER deployments do not scale automatically. For more information, refer to [ELSER](../explore-analyze/machine-learning/nlp/ml-nlp-elser.md) and [Trained model autoscaling](../explore-analyze/machine-learning/nlp/ml-nlp-auto-scale.md).
> :::
> 
> :::{tab-item} {{ece}}
> * In the event that an override is set for the instance size or disk quota multiplier for an instance by means of the [Instance Overrides API](https://www.elastic.co/docs/api/doc/cloud-enterprise/operation/operation-set-all-instances-settings-overrides), autoscaling will be effectively disabled. It’s recommended to avoid adjusting the instance size or disk quota multiplier for an instance that uses autoscaling, since the setting prevents autoscaling.
> :::
> 
> ::::

Reviewer: we generally want to only use tabs when we have two parallel paths to outline: one for context x and one for context y.

Because these are not parallel paths, we don't want to use tabs. we probably want to use prose instead:

```suggestion

In {{ech}} the following additional limitations apply:

* Trial deployments cannot be configured to autoscale beyond the normal Trial deployment size limits. The maximum size per zone is increased automatically from the Trial limit when you convert to a paid subscription.
* ELSER deployments do not scale automatically. For more information, refer to [ELSER](../explore-analyze/machine-learning/nlp/ml-nlp-elser.md) and [Trained model autoscaling](../explore-analyze/machine-learning/nlp/ml-nlp-auto-scale.md).

In {{ece}}, the following additional limitations apply:

* In the event that an override is set for the instance size or disk quota multiplier for an instance by means of the [Instance Overrides API](https://www.elastic.co/docs/api/doc/cloud-enterprise/operation/operation-set-all-instances-settings-overrides), autoscaling will be effectively disabled. It’s recommended to avoid adjusting the instance size or disk quota multiplier for an instance that uses autoscaling, since the setting prevents autoscaling.
```

## Enable or disable autoscaling[ec-autoscaling-enable]

$$$ece-autoscaling-enable$$$$$$ech-autoscaling-enable$$$To enable or disable autoscaling on a deployment:

> 1. **Log in** to the console.
> 
> ::::{tab-set}
> 
> :::{tab-item} {{ech}}
> [Elasticsearch Service Console](https://cloud.elastic.co?page=docs&placement=docs-body)
> :::
> 
> :::{tab-item} {{ece}}
> [Cloud UI](../deploy-manage/deploy/cloud-enterprise/log-into-cloud-ui.md)
> :::
> 
> :::{tab-item} {{ecloud}} -  Heroku
> [Elasticsearch Add-On for Heroku console](https://cloud.elastic.co?page=docs&placement=docs-body)
> :::
> ::::

Reviewer: we need to land on a generic term, but let's just squish them together for now. no need to provide a heroku option.

we no longer use the term "Elasticsearch Service" - only "Elastic Cloud" or "Elastic Cloud Hosted" depending on whether you're referring to the organization or the deployment.

```suggestion
1. Log in to the ECE [Cloud UI]((/deploy-manage/deploy/cloud-enterprise/log-into-cloud-ui.md) or  [{{ecloud}} Console](https://cloud.elastic.co?page=docs&placement=docs-body).
```

2. On the **Deployments** page, select your deployment.

    Narrow your deployments by name, ID, or choose from several other filters. To customize your view, use a combination of filters, or change the format from a grid to a list.

3. In your deployment menu, select **Edit**.
4. Select desired autoscaling configuration for this deployment using **Enable Autoscaling for:** dropdown menu.
5. Select **Confirm** to have the autoscaling change and any other settings take effect. All plan changes are shown on the Deployment **Activity** page.

When autoscaling has been enabled, the autoscaled nodes resize according to the [autoscaling settings](../deploy-manage/autoscaling.md#ec-autoscaling-update). Current sizes are shown on the deployment overview page.

When autoscaling has been disabled, you need to adjust the size of data tiers and machine learning nodes manually.

## Update your autoscaling settings[ec-autoscaling-update]

$$$ece-autoscaling-update$$$$$$ech-autoscaling-update$$$Each autoscaling setting is configured with a default value. You can adjust these if necessary, as follows:

1. **Log in** to the console.

> ::::{tab-set}

Reviewer: same here
